source/extensions/common/wasm/wasm_vm.cc

```diff
@@ -0,0 +1,31 @@
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include <memory>
+
+#include "extensions/common/wasm/null/null.h"
+#include "extensions/common/wasm/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+thread_local Envoy::Extensions::Common::Wasm::Context* current_context_ = nullptr;
+thread_local uint32_t effective_context_id_ = 0;
+
+WasmVmPtr createWasmVm(absl::string_view wasm_vm) {
+  if (wasm_vm.empty()) {
+    throw WasmException("Failed to create WASM VM with unspecified runtime.");
+  } else if (wasm_vm == WasmVmNames::get().Null) {
+    return Null::createVm();
```

Comment:
What is the purpose of the null VM? Should this just go in the test directory? Along these lines do we need a registration system for WASM VMs similar to other envoy registrations?

Reply:
See [my previous comment](https://github.com/envoyproxy/envoy/pull/8020#pullrequestreview-278766766):
> Null VM is the shim implementation of Wasm VM, which allows code written using Wasm SDK/API to be compiled as native Envoy extensions, instead of WebAssembly modules.

Reply:
The Null VM allows code to be compiled directly into Envoy using the WASM APIs.  This is nice for testing because it means that gdb for example works, e.g. breakpoints can be set it the plugin code.  It is also a good way to measure the overhead of the VM system itself for regression and tuning.  This is how istio is currently being developed.   The code can then just be recompiled and loaded up as a WASM module.

We could use the registration system for the WASM VMs.   There are only 3 currently (null, v8, wavm), so I didn't bother, but if you like I can change it.

We are using the registration system for Null VM plugin.

Reply:
No it's fine if the null VM has actual use we can leave it here as is.